hail/python/hail/methods/statgen.py

```diff
@@ -440,49 +441,86 @@ def select_array_indices(hl_array, indices):
-    ys_and_covs_to_keep = ys_and_covs_to_keep_with_indices.map(lambda pair: pair[1])
+    list_of_ys_and_covs_to_keep_with_indices = \
+        [hl.enumerate(ht[sample_field_name]).filter(lambda struct_with_index: all_defined(struct_with_index[1], one_y_field_name_set + cov_field_names)) for one_y_field_name_set in y_field_names]
+
+    def make_one_cov_matrix(ys_and_covs_to_keep):
+        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, cov_field_names))) \
+            if cov_field_names else hl.nd.zeros((hl.len(ys_and_covs_to_keep), 0))
+
+    def make_one_y_matrix(ys_and_covs_to_keep, one_y_field_name_set):
+        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, one_y_field_name_set)))
+
+    list_of_ys_and_covs_to_keep = [inner_list.map(lambda pair: pair[1]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
+    list_of_indices_to_keep = [inner_list.map(lambda pair: pair[0]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
+
+    cov_nds = [make_one_cov_matrix(ys_and_covs_to_keep) for ys_and_covs_to_keep in list_of_ys_and_covs_to_keep]
+
+    y_nds = [make_one_y_matrix(ys_and_covs_to_keep, one_y_field_name_set)
+             for ys_and_covs_to_keep, one_y_field_name_set in zip(list_of_ys_and_covs_to_keep, y_field_names)]
 
-    cov_nd = hl.nd.array(ys_and_covs_to_keep.map(lambda struct: hl.array([struct[cov_name] for cov_name in cov_field_names]))) if cov_field_names else hl.nd.zeros((hl.len(indices_to_keep), 0))
-    ht = ht.annotate_globals(kept_samples=indices_to_keep,
-                             __y_nd=hl.nd.array(ys_and_covs_to_keep.map(lambda struct: hl.array([struct[y_name] for y_name in y_field_names]))),
-                             __cov_nd=cov_nd)
+    ht = ht.annotate_globals(kept_samples=list_of_indices_to_keep,
+                             __y_nds=y_nds,
+                             __cov_nds=cov_nds)
     k = builtins.len(covariates)
-    n = hl.len(ht.kept_samples)
-    ht = ht.annotate_globals(d=n - k - 1)
-    cov_Qt = hl.if_else(k > 0, hl.nd.qr(ht.__cov_nd)[0].T, hl.nd.zeros((0, n)))
-    ht = ht.annotate_globals(__Qty=cov_Qt @ ht.__y_nd)
-    ht = ht.annotate_globals(__yyp=dot_rows_with_themselves(ht.__y_nd.T) - dot_rows_with_themselves(ht.__Qty.T))
+    ns = ht.index_globals().kept_samples.map(lambda one_sample_set: hl.len(one_sample_set))
+    ht = ht.annotate_globals(ds=ns.map(lambda n: n - k - 1))
+    cov_Qts = hl.if_else(k > 0,
+                         ht.__cov_nds.map(lambda one_cov_nd: hl.nd.qr(one_cov_nd)[0].T),
+                         ns.map(lambda n: hl.nd.zeros((0, n))))
+    ht = ht.annotate_globals(__cov_Qts=cov_Qts)
+    ht = ht.annotate_globals(__Qtys=hl.range(num_y_lists).map(lambda i: ht.__cov_Qts[i] @ ht.__y_nds[i]))
+    ht = ht.annotate_globals(__yyps=hl.range(num_y_lists).map(lambda i: dot_rows_with_themselves(ht.__y_nds[i].T) - dot_rows_with_themselves(ht.__Qtys[i].T)))
```

Comment:
I think this chain of `annotate_globals` can be cleaned up, making use of the python CSE. Something like
```
cov_Qts = hl.if_else(k > 0,
                     cov_nds.map(lambda one_cov_nd: hl.nd.qr(one_cov_nd)[0].T),
                     ns.map(lambda n: hl.nd.zeros((0, n))))
Qtys = hl.range(num_y_lists).map(lambda i: cov_Qts[i] @ y_nds[i])
ht = ht.annotate_globals(
  kept_samples=list_of_indices_to_keep,
  __y_nds=y_nds,
  __cov_nds=cov_nds,
  ds=ns.map(lambda n: n - k - 1),
  __cov_Qts=cov_Qts,
  __Qtys=Qtys,
  __yyps=hl.range(num_y_lists).map(lambda i: dot_rows_with_themselves(y_nds[i].T) - dot_rows_with_themselves(Qtys[i].T))
```
(possibly dropping some of those from the `annotate_globals` if they aren't actually needed again.

Reply:
I did this and was able to delete `__cov_nds`, so that was nice. `kept_samples` currently needs to stay where it is due to what I believe is a bug where we are losing the `source` of an expression's `_indices`. I have yet to discover why.